#4. Регистрация/Логин

Neyrzx opened the pull request:
Реализация системы аутентификации и авторизации (#4)

### Что сделано ✅
#### Основной функционал
1. Реализована система аутентификации:
   - Регистрация пользователя с валидацией:
     * Проверка формата email
     * Политика сложности паролей
     * Проверка уникальности email
   - Логин с выдачей JWT токенов
   - Обновление токенов через refresh token
   - Middleware для проверки авторизации

2. Безопасность:
   - Хеширование паролей с bcrypt
   - Использование ECDSA для подписи JWT
   - Защита от повторного использования токенов

### Что не хватает ⚠️

1. Тестирование:
   - Юнит-тесты для сервисов
   - Интеграционные тесты API
   - Тесты безопасности

2. Доработки безопасности:
   - Реализация MX-проверки email (сейчас отключено в конфиге)
   - Механизм блокировки при brute-force атаках (rate limit)
   - Валидация fingerprint устройства (в моделях есть, но не реализованно в сервисах)

3. Документация:
   - Swagger-спека.

4. Конфигурирование:
   - Вынесение параметров токенов (время жизни) в конфиг
   - Настройка секретов через environment variables

### Хорошо бы задачи на 🚀

1. Дополнительные функции:
   - Подтверждение email
   - Восстановление пароля
   - OAuth-интеграция!!

2. Оптимизации:
   - Кеширование активных сессий
   - Фоновое удаление устаревших сессий
   - Rate-limit-ер на эндпоинты аутентификации

### Примеры запросов
   
```http
###
POST http://localhost:8080/api/auth/register HTTP/1.1
Content-Type: application/json

{
    "username": "username9",
    "email": "username9@email.com",
    "password": "hardpassword6Q"
}

###
POST http://localhost:8080/api/auth/login HTTP/1.1
Content-Type: application/json

{
    "email": "username9@email.com",
    "password": "hardpassword6Q"
}

###
POST http://localhost:8080/api/auth/refresh HTTP/1.1
Refresh-Token: eyJhbGciOiJFUzI1NiIsInR5cCI6IkpXVCJ9.eyJ1aWQiOjEsInJvbCI6MCwidHlw...


###
POST http://localhost:8080/api/auth/refresh HTTP/1.1

{
    "refresh_token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdHJhd2JlcnJ5Iiwi..."
}
```

Всё ещё нужны тесты.

PosokhovVadim commented on internal/repository/user.go:
```diff
@@ -0,0 +1,58 @@
+
+// Insert - сохраняет пользователя в бд.
+func (r *UserRepository) Insert(ctx context.Context, dto auth.RegisterUser) (id uint, err error) {
+	userModel := model.ConvertUserFromSvc(dto)
```
Предлагаю перенести конверт данных на уровень сервиса, так чтобы методы репозитория принимали модель репы.

Neyrzx replied:
По идеи все конвертации доменных сущностей проходят на внешних слоях.
Делается это для того, чтобы наш доменный слой ничего не знал о моделях внешних слоёв, и зависел ток от своих доменных сущностей. Тип полная изоляция, дает возможность без изменения доменного слоя менять реализации внешних слоев, контроллеров, гейтвеев и т.п.

Типо все знают о доменных сушностях, но доменный слой ничего не знает о внешних, и рабоатет только через свои.

Вроде у нас чистая архитектура, кажется так как оно сейчас реализовано больше под неё подходит.

Neyrzx committed:
feat(api): add authentication endpoints (#4)

- Implement registration, login and refresh token endpoints
- Create auth error handling infrastructure
- Integrate auth routes into main router
- Add DTOs for auth requests/responses

Neyrzx committed:
feat(session): implement session management service (#4)

- Add JWT token generation with ECDSA signing
- Implement refresh token rotation mechanism
- Add session repository interface and error types
- Integrate with bcrypt password hasher
- Add session validation with device fingerprint checks
- Implement token verification with expiration handling

Neyrzx committed:
feat(auth): integrate authentication components with main app (#4)

- Add session and user models to database migrations
- Implement user and session repository implementations
- Wire up auth service dependencies in main entrypoint
- Configure bcrypt hasher with minimal cost for development
- Setup transaction support for database operations
- Add model conversion utilities between layers

Neyrzx committed:
feat(auth): implement JWT authentication middleware

- Add bearer token validation with session manager
- Implement proper authorization header parsing
- Handle token expiration and invalid token cases
- Integrate with session service for token verification
- Add structured error responses for auth failures

Neyrzx committed:
chore(deps): add authentication and security dependencies

- Add JWT v5 library for token handling
- Include UUID package for session identifiers
- Update crypto and net packages for security
- Sync dependency versions in go.sum
